File: apps/android/app/src/main/java/com/exponential/app/ui/components/GlassComposer.kt
package com.exponential.app.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.exponential.app.ui.theme.DesignTokens
import com.exponential.app.ui.theme.GlassTokens
import com.exponential.app.ui.theme.TextEmphasis

/**
 * The ONE composer shell (EXP-698) — the issue detail's docked comment box,
 * the steer screen's "message the agent" box and the support thread's reply
 * box. All three drew the same thing by hand with different corner radii (24
 * / 24 / none), different fills and a different tool row; they are this now.
 *
 * The card is the standard [GlassTokens.CardRadius] glass card, NOT a 24dp
 * capsule-ish blob: a composer is a card with a field in it, the same rung as
 * every other card on the screen. [opaque] is for the two that FLOAT over
 * scrolling content (the docked comment bar, the steer bar) — the translucent
 * fill lets the feed ghost through otherwise. No elevation on either: the
 * hairline is the edge.
 *
 * Slots, top to bottom: [leading] (a row above the field — the support
 * thread's Reply / Internal-note pills), [strip] (queued attachments),
 * [field], and a bottom row of [tools] (ghost glyph buttons) with [submit]
 * pushed to the end.
 */
@Composable
fun GlassComposer(
    modifier: Modifier = Modifier,
    opaque: Boolean = false,
    leading: (@Composable ColumnScope.() -> Unit)? = null,
    strip: (@Composable ColumnScope.() -> Unit)? = null,
    tools: (@Composable RowScope.() -> Unit)? = null,
    submit: (@Composable () -> Unit)? = null,
    field: @Composable ColumnScope.() -> Unit,
) {
    val shape = RoundedCornerShape(GlassTokens.CardRadius)
    Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(shape)
            .background(if (opaque) GlassTokens.OpaqueCardFill else GlassTokens.CardFill, shape)
            .border(
                GlassTokens.Hairline,
                if (opaque) GlassTokens.StrokeStrong else GlassTokens.StrokeCard,
                shape,
            )
            .padding(
                horizontal = GlassComposerDefaults.HorizontalPadding,
                vertical = GlassComposerDefaults.VerticalPadding,
            ),
    ) {
        leading?.invoke(this)
        strip?.invoke(this)
        field()
        if (tools != null || submit != null) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                tools?.invoke(this)
                Spacer(Modifier.weight(1f))
                submit?.invoke()
            }
        }
    }
}

/**
 * One ghost glyph in a composer's tool row — attach, mention, issue-ref,
 * emoji. No chrome of its own (the composer card is the surface); a secondary
 * glyph that dims to quaternary when it cannot act.
 */
@Composable
fun ComposerToolButton(
    icon: ImageVector,
    contentDescription: String,
    onClick: () -> Unit,
    enabled: Boolean = true,
) {
    IconButton(onClick = onClick, enabled = enabled) {
        Icon(
            icon,
            contentDescription = contentDescription,
            modifier = Modifier.size(GlassComposerDefaults.ToolGlyphSize),
            tint = Color.White.copy(
                alpha = if (enabled) TextEmphasis.Secondary else TextEmphasis.Quaternary,
            ),
        )
    }
}

/**
 * The composer's submit — a ghost round button whose GLYPH carries the state:
 * full white when there is something to send, quaternary when there is not,
 * a spinner while it is in flight. No filled capsule: the send arrow is the
 * only bright thing in the composer, so it needs no second signal.
 */
@Composable
fun ComposerSubmitButton(
    icon: ImageVector,
    contentDescription: String,
    onClick: () -> Unit,
    enabled: Boolean,
    sending: Boolean = false,
) {
    IconButton(onClick = onClick, enabled = enabled && !sending) {
        if (sending) {
            CircularProgressIndicator(
                modifier = Modifier.size(GlassComposerDefaults.SpinnerSize),
                strokeWidth = 2.dp,
                color = Color.White,
            )
        } else {
            Icon(
                icon,
                contentDescription = contentDescription,
                modifier = Modifier.size(GlassComposerDefaults.SubmitGlyphSize),
                tint = if (enabled) Color.White else Color.White.copy(alpha = TextEmphasis.Quaternary),
            )
        }
    }
}

/** The composer's own numbers, so no call site re-types an inset (EXP-698). */
object GlassComposerDefaults {
    val HorizontalPadding: Dp = 12.dp
    val VerticalPadding: Dp = 8.dp
    val ToolGlyphSize: Dp = DesignTokens.Size.ControlSm
    val SubmitGlyphSize: Dp = 24.dp
    val SpinnerSize: Dp = 18.dp
}

File: apps/ios/ExpUI/Sources/GlassComposer.swift
import SwiftUI

// The ONE composer card (EXP-698 round 2). iOS grew three of them — the issue
// comment composer, the steer composer and the support reply box — and they
// had drifted onto three different chromes: one `opaqueCardFill` at radius 20
// with `strokeStrong`, one `.ultraThinMaterial` at radius 20 with a hand-typed
// `white.opacity(0.12)` hairline, and one `white.opacity(0.06)` at radius 18
// inside an `.ultraThinMaterial` bar. All three carried the same drop shadow,
// the one thing the flat glass language does not have.
//
// `GlassComposer` is that card: `Radius.xl` corners, `fillCard` + the
// `strokeCard` hairline, no material and no shadow. `isOpaque` swaps in
// `opaqueCardFill` + `strokeStrong` for a composer that FLOATS over scrolling
// content, the same switch `GlassCard` and `GlassPill` carry.
//
// The card owns the chrome and the tool row; the slots keep their own inner
// padding, because a `MarkdownEditor` pads itself and a bare `TextField` does
// not. Slot ORDER is leading · field · strip · tools: the attachment strip
// sits BELOW the text and above the action row on every client, so a queued
// image reads as something the message you are typing carries, not as a
// banner over it.

/// The pinned geometry of a composer card, so the three composers cannot drift
/// apart again.
public enum GlassComposerTokens {
    public static let cornerRadius: CGFloat = DesignTokens.Radius.xl
    /// The glyph inside a tool button.
    public static let toolGlyphSize: CGFloat = AppIcon.Size.medium
    /// A tool button's TAP TARGET. Deliberately much larger than the glyph:
    /// these sit in a row of five, a thumb-width apart, above the keyboard —
    /// sizing the frame to the 17pt mark would leave five 17pt targets and a
    /// composer you cannot hit. The glyphs still read as small ghost marks.
    public static let toolHitSize: CGFloat = 36
    /// The submit glyph — a round mark, drawn ghost (no capsule behind it).
    public static let submitSize: CGFloat = 28
    /// Gap between two tool buttons — tight, because `toolHitSize` already
    /// carries the visual pitch.
    public static let toolSpacing: CGFloat = 2
    public static let horizontalPadding: CGFloat = 12
    public static let bottomPadding: CGFloat = 10
}

// MARK: - Tool + submit buttons

/// A ghost glyph button in a composer's tool row (photo, attach, @, #, emoji).
public struct GlassComposerToolButton: View {
    let icon: String
    let accessibilityLabel: String
    var enabled: Bool = true
    let action: () -> Void

    public init(
        _ icon: String,
        accessibilityLabel: String,
        enabled: Bool = true,
        action: @escaping () -> Void
    ) {
        self.icon = icon
        self.accessibilityLabel = accessibilityLabel
        self.enabled = enabled
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            AppIcon(icon, size: GlassComposerTokens.toolGlyphSize)
                .foregroundStyle(
                    .white.opacity(enabled ? TextOpacity.secondary : TextOpacity.quaternary)
                )
                .frame(
                    width: GlassComposerTokens.toolHitSize,
                    height: GlassComposerTokens.toolHitSize
                )
                .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .disabled(!enabled)
        .accessibilityLabel(accessibilityLabel)
    }
}

/// The round ghost submit at the trailing end of a composer's tool row.
/// `tint` is the LIVE color (white by default; the support composer tints an
/// internal note); disabled drops to the quaternary white.
public struct GlassComposerSubmitButton: View {
    let icon: String
    let accessibilityLabel: String
    var enabled: Bool = true
    var tint: Color = .white
    let action: () -> Void

    public init(
        _ icon: String,
        accessibilityLabel: String,
        enabled: Bool = true,
        tint: Color = .white,
        action: @escaping () -> Void
    ) {
        self.icon = icon
        self.accessibilityLabel = accessibilityLabel
        self.enabled = enabled
        self.tint = tint
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            AppIcon(icon, size: GlassComposerTokens.submitSize)
                .foregroundStyle(enabled ? tint : .white.opacity(TextOpacity.quaternary))
                .contentShape(Circle())
        }
        .buttonStyle(.plain)
        .disabled(!enabled)
        .accessibilityLabel(accessibilityLabel)
    }
}

// MARK: - The card

public struct GlassComposer<
    Leading: View,
    Strip: View,
    Field: View,
    Tools: View,
    Submit: View
>: View {
    /// Lays the opaque card surface beneath the glass tint — for a composer
    /// that floats over a scrolling feed.
    var isOpaque: Bool = false
    let leading: Leading
    let strip: Strip
    let field: Field
    let tools: Tools
    let submit: Submit

    /// Slots are declared in the order they RENDER.
    public init(
        isOpaque: Bool = false,
        @ViewBuilder leading: () -> Leading,
        @ViewBuilder field: () -> Field,
        @ViewBuilder strip: () -> Strip,
        @ViewBuilder tools: () -> Tools,
        @ViewBuilder submit: () -> Submit
    ) {
        self.isOpaque = isOpaque
        self.leading = leading()
        self.field = field()
        self.strip = strip()
        self.tools = tools()
        self.submit = submit()
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            leading
            field
            strip
            HStack(spacing: GlassComposerTokens.toolSpacing) {
                tools
                Spacer(minLength: 0)
                submit
            }
            .padding(.horizontal, GlassComposerTokens.horizontalPadding)
            .padding(.bottom, GlassComposerTokens.bottomPadding)
        }
        .background(isOpaque ? GlassTokens.opaqueCardFill : GlassTokens.fillCard)
        .clipShape(RoundedRectangle(cornerRadius: GlassComposerTokens.cornerRadius))
        .overlay(
            RoundedRectangle(cornerRadius: GlassComposerTokens.cornerRadius)
                .stroke(
                    isOpaque ? GlassTokens.strokeStrong : GlassTokens.strokeCard,
                    lineWidth: GlassTokens.hairline
                )
        )
    }
}

// MARK: - Slot-omitting inits

extension GlassComposer where Leading == EmptyView {
    /// No leading row — the issue and steer composers.
    public init(
        isOpaque: Bool = false,
        @ViewBuilder field: () -> Field,
        @ViewBuilder strip: () -> Strip,
        @ViewBuilder tools: () -> Tools,
        @ViewBuilder submit: () -> Submit
    ) {
        self.init(
            isOpaque: isOpaque,
            leading: { EmptyView() },
            field: field,
            strip: strip,
            tools: tools,
            submit: submit
        )
    }
}

extension GlassComposer where Strip == EmptyView, Tools == EmptyView {
    /// A leading row and no tools — the support reply box.
    public init(
        isOpaque: Bool = false,
        @ViewBuilder leading: () -> Leading,
        @ViewBuilder field: () -> Field,
        @ViewBuilder submit: () -> Submit
    ) {
        self.init(
            isOpaque: isOpaque,
            leading: leading,
            field: field,
            strip: { EmptyView() },
            tools: { EmptyView() },
            submit: submit
        )
    }
}
